src/Psalm/Internal/Codebase/Methods.php

```diff
@@ -686,21 +694,55 @@ public function getMethodReturnType(
         $appearing_fq_class_storage = $this->classlike_storage_provider->get($appearing_fq_class_name);
 
         if ($appearing_fq_class_name === 'UnitEnum'
-            && $original_method_name === 'cases'
             && $original_class_storage->is_enum
-            && $original_class_storage->enum_cases
         ) {
-            $types = [];
+            if ($original_method_name === 'cases') {
+                if ($original_class_storage->enum_cases === []) {
+                    return Type::getEmptyArray();
+                }
+                $types = [];
 
-            foreach ($original_class_storage->enum_cases as $case_name => $_) {
-                $types[] = new Type\Union([new Type\Atomic\TEnumCase($original_fq_class_name, $case_name)]);
-            }
+                foreach ($original_class_storage->enum_cases as $case_name => $_) {
+                    $types[] = new Type\Union([new Type\Atomic\TEnumCase($original_fq_class_name, $case_name)]);
+                }
 
-            $list = new Type\Atomic\TKeyedArray($types);
-            $list->is_list = true;
-            $list->sealed = true;
+                $list = new Type\Atomic\TKeyedArray($types);
+                $list->is_list = true;
+                $list->sealed = true;
+                return new Type\Union([$list]);
+            }
+        }
 
-            return new Type\Union([$list]);
+        if ($appearing_fq_class_name === 'BackedEnum'
+            && $original_class_storage->is_enum
+            && $original_class_storage->enum_type
+        ) {
+            if (($original_method_name === 'from'
+                || $original_method_name === 'tryfrom'
+                ) && $source_analyzer
+                && isset($args[0])
+                && ($first_arg_type = $source_analyzer->getNodeTypeProvider()->getType($args[0]->value))
+            ) {
+                $types = [];
+                foreach ($original_class_storage->enum_cases as $case_name => $case_storage) {
+                    if (UnionTypeComparator::isContainedBy(
+                        $source_analyzer->getCodebase(),
+                        \is_int($case_storage->value) ?
```

Comment:
Is it possible to use a constant to initialize a case? If it happens, is_int() will fail. Can't we access the type at this moment?

Reply:
It's possible, and it fails. But not here.

https://3v4l.org/WHXIV#v8.1rc3
https://psalm.dev/r/3547ca91d3

Reply:
Built-in constants work. 
User-defined constants don't: https://3v4l.org/nhGRA#v8.1rc3
Class constants work if the class is defined before the enum: https://3v4l.org/bcYPN#v8.1rc3.
Class constants from autoloaded classes do not work: https://3v4l.org/AjOCi#v8.1rc3

3v4l doesn't have proper 8.1 yet, so it's based on RC3. Some of the above could have been fixed before final release.

Reply:
> Some of the above could have been fixed before final release.

I've just built 8.1 branch, and there are no changes for those cases.